base/utils/R/combine_model_netcdf.R

```diff
@@ -0,0 +1,167 @@
+  # loop over time.
+  # initialize parallel.
+  cl <- parallel::makeCluster(as.numeric(cores))
+  doSNOW::registerDoSNOW(cl)
+  #progress bar
+  pb <- utils::txtProgressBar(min = 1, max = length(site_info$site_id), style = 3)
+  progress <- function(n) utils::setTxtProgressBar(pb, n)
+  opts <- list(progress=progress)
+  # record nc paths.
+  nc.paths <- c()
+  for (t in seq_along(time.points)) {
+    time <- time.points[t] # grab the current time point.
+    # loop over sites.
+    nc.files <- 
+      foreach::foreach(s = seq_along(site_info$site_id), 
+                       .packages = c("Kendall", "purrr", "ncdf4"), 
+                       .options.snow=opts) %dopar% {
+                         single_site_nc_merge(model.outdir = model.outdir, 
+                                              nc.outdir = nc.outdir, 
+                                              ens.num = ens.num, 
+                                              # cdo collgrid only works for numeric data type.
+                                              site.id = as.numeric(site_info$site_id[s]), 
+                                              lat = site_info$lat[s], 
+                                              lon = site_info$lon[s], 
+                                              time)
+                       } %>% unlist
+    # merge across sites using CDO command.
+    cmd <- "cdo -P @CORES@ collgrid @NC.OUTDIR@/*.nc @OUTFILE@"
+    cmd <- gsub("@CORES@", cores, cmd)
+    cmd <- gsub("@NC.OUTDIR@", nc.outdir, cmd)
+    cmd <- gsub("@OUTFILE@", file.path(nc.outdir, paste0(time, ".nc")), cmd)
+    out <- system(cmd, intern = TRUE)
+    # record the current nc path.
+    nc.paths <- c(nc.paths, file.path(nc.outdir, paste0(time, ".nc")))
+    # remove nc files for each site.
+    unlink(nc.files)
+  }
+  # stop parallel.
+  parallel::stopCluster(cl)
+  foreach::registerDoSEQ()
+  # return nc paths.
+  return(nc.paths)
+}
+
+#' Merge model outputted netCDF files across ensembles for a single site.
+#' @details
+#' The function is only tested for SIPNET model runs.
+#' Please make sure you have the same netCDF formats if you want to proceed with different models.
+#' 
+#' @param model.outdir character: physical path to the model output folder.
+#' @param nc.outdir  character: physical path to the folder that contains the merged netCDF files.
+#' @param ens.num numeric: ensemble size.
+#' @param site.id numeric: identification number of the site.
+#' @param lat numeric: latitude of the site.
+#' @param lon numeric: longitude of the site.
+#' @return character: file path to the merged netCDF file.
+#' @export
+#' 
+#' @author Dongchen Zhang
+single_site_nc_merge <- function (model.outdir, nc.outdir, ens.num, site.id, lat, lon, time) {
+  # grab basic formats from the first nc file of the site.
+  # create the folder name associated with first ensemble and first site.
+  prefix <- "ENS-"
+  folder.name <- paste0(prefix, sprintf("%05d", 1), "-", 1)
```

Comment:
Shouldn't the folder name have the site id in it?
```suggestion
  folder.name <- paste0(prefix, sprintf("%05d", 1), "-", site.id)
```

Reply:
Or: rather than construct names from ensemble numbers, consider something like `folders <- list.files(out.dir, pattern = site.id, recursive = FALSE)` and then iterate over those?

Reply:
If helpful, I used the runs.txt file to build the output directories:

https://github.com/ccmmf/downscaling/blob/dd4235cd5b6dd8d59d46f128a560b2fbf240981e/scripts/030_extract_sipnet_output.R#L32-L43

```
# Read the runs.txt file to get ensemble and site and directory information
runs_file <- file.path(pecan_outdir, "run", "runs.txt")
## Expected outputs based on runs.txt
expected_ens_dirs <- tibble::tibble(
    dir = readr::read_lines(runs_file)
) |>
    tidyr::separate(
        col = "dir",
        into = c("prefix", "ens", "site_id"),
        sep = "-",
        remove = FALSE
    )
```

Reply:
I prefer to create the `folder.name` from `ens` and `site_id` rather than do it inversely, because that's what we have from the initial. And I also trust the naming convention more than `list.files`.

Reply:
Those are fair points, but do make sure you actually use `site.id` instead of the current hardcoded `1`.

(I may want to revisit this later as part of an effort to move the behavior from "enforce a predetermined run design" more toward "here, merge everything in this directory unless the dimensions are incompatible", but that can come later)